### ISSUE TITLE
Subchannel connection has seemingly arbitrary deadline

### ISSUE THREAD

**Issue by alanctgardner** [grpc/grpc#12830]
### What version of gRPC and what language are you using?

Github master, affects C core library. 
 
### What operating system (Linux, Windows, …) and version?
 
I'm on OS X and Linux, affects all platforms.
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
clang - `Apple LLVM version 8.1.0 (clang-802.0.42)`
python - `2.7.13`
ruby - `ruby 2.3.3p222`
 
### What did you do?

- set the configuration options `grpc.min_reconnect_backoff_ms`, `grpc.max_reconnect_backoff_ms`, `grpc.initial_reconnect_backoff_ms`.

- attempt to make an RPC like normal

### What did you expect to see?

Expected gRPC to attempt to open a connection. If that connection failed, it would wait for `grpc.initial_reconnect_backoff_ms` milliseconds to attempt reconnecting. Afterwards it would attempt to reconnect at jittered intervals between `grpc.min_reconnect_backoff_ms` and `grpc.max_reconnect_backoff_ms`
 
### What did you see instead?
 
`continue_connect_locked` uses the reconnect backoff value as a deadline! Instead of trying to reconnect quickly after a connection failure, the subchannel would fail to handshake because the connection would timeout. This is totally the opposite of the expected behaviour. You can already set a deadline for the entire RPC, I don't understand the value of setting a deadline for the handshake. At the very least this should be named more clearly or the timeout should be separate from the backoff.
(https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/subchannel.c#L381-L394)

Setting the deadline to `c->next_attempt` doesn't make any sense to me in this context.

Sample logs:
This is a sample error showing the handshaker timing out when the initial backoff is set very low.

```
D1004 16:46:59.775395000 123145356509184 client_channel.c:1331]        chand=0x7fbfba809dd0 calld=0x7fbfb80d8f20: failing batch with error: {"created":"@1507150015.843660000","description":"Failed to create subchannel","file":"src/core/ext/filters/client_channel/client_channel.c","file_line":1033,"referenced_errors":[{"created":"@1507150015.843404000","descrip
tion":"Pick Cancelled","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c","file_line":170,"referenced_errors":[{"created":"@1507150015.843015000","description":"Connect Failed","file":"src/core/ext/filters/client_channel/subchannel.c","file_line":684,"grpc_status":14,"referenced_errors":[{"created":"@1507150015.842909000","descripti
on":"Handshake read failed","file":"src/core/lib/security/transport/security_handshaker.c","file_line":308,"referenced_errors":[{"created":"@1507150015.842798000","description":"FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.c","file_line":466,"referenced_errors":[{"created":"@1507150015.842755000","description":"Handshake timed out","file":"src/core/li
b/channel/handshaker.c","file_line":227}]}]}]}]}]}
```
 
### Anything else we should know about your project / environment?

Are you open to a PR to set the deadline for subchannel connection to unlimited?

**Comment by markdroth** [comment]
Duplicate of #12867.

**Comment by markdroth** [comment]
I take back what I said about; I think this is actually a separate issue from #12867.

We definitely cannot eliminate the deadline for handshaking.  There needs to be one, or else there is the possibility of DoS attacks by clients establishing many TCP connections and then never sending any subsequent packets.  However, I am open to the possibility of setting the deadline for this to some other value, rather than using the time of the next backoff reconnect attempt.

@dgquintas and/or @ctiller, do either of you have any historical knowledge about why we use the next reconnection attempt time as the deadline for handshaking?  I guess it sort of makes sense if we consider that the back-off applies to the entire connection establishment process, which includes both establishing the TCP connection and doing any necessary handshaking.  But it does seem like we should not give up on establishing a connection if we are actually making forward progress (e.g., we have established the TCP connection and are in the process of handshaking).

@ejona86 @dfawley as an FYI about the DoS issues we've been discussing.

**Comment by markdroth** [comment]
Actually, further thought: This may not be a DoS attack issue, since this is the client side, not the server side.  I need to check where we set this deadline on the server side, but that's a separate issue.

Still, I do think a deadline on the client side is also valuable, although as I said above, it may make sense to set it to something unrelated to the reconnection backoff value.

**Comment by alanctgardner** [comment]
> I guess it sort of makes sense if we consider that the back-off applies to the entire connection establishment process, which includes both establishing the TCP connection and doing any necessary handshaking.

To provide an example of how I'm thinking about backoff, look at [gax-go](https://github.com/googleapis/gax-go/blob/master/call_option.go#L83-L95) - the backoff retrier:
- issues the call (which may have an associated timeout)
- the call returns an error
- the backoff retrier starts the backoff window
- the retrier makes another call after the backoff window has elapsed

By contrast this code:
- Starts the backoff window
- Starts the handshake with a deadline equal to the backoff window
- Cancels the handshake when the deadline is exceeded
- Immediately retries the handshake since the backoff window has already elapsed

I think it's pretty clearly a bug to conflate the deadline for the handshake to complete, and the backoff, which is by definition the time between failed handshakes to wait before retrying. I think it would be appropriate to have two different options, one to control the handshake timeout, and a second to control the interval between handshakes when they need to be retried.

**Comment by dfawley** [comment]
FWIW, Go uses a 20 second minimum timeout for client-side connection establishment, and goes up from there as the backoff grows.  I am not familiar with the history/rationale of that implementation.

Relevant code:
https://github.com/grpc/grpc-go/blob/5856538706dc3abc44dd5ba1a8bb19a0dfb1ab25/clientconn.go#L918

**Comment by markdroth** [comment]
Doug, the connection backoff spec for establishing a TCP connection is here:

https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md

The question in this particular bug is what deadline to use for whatever handshaking has to be done after the TCP connection has been established.

**Comment by alanctgardner** [comment]
Reading that pseudocode made it clearer to me how this parameter is intended to work - you set a deadline for the connection to succeed, and if it times out you immediately retry. In other failure modes you may end up sleeping for some time during the backoff, but in the timeout case the timeout pulls double duty as the backoff time. There is no way to have a timeout longer than the backoff interval because you would end up with multiple concurrent connection attempts.

To my earlier point, I don't know that it's a very common definition of "backoff" to include the time spent in the call you're backing off from. My naive interpretation of how a backoff parameter should work is:

```
ConnectWithBackoff()
  // Potentially accept some user-specified timeout for reconnecting, which must be > MIN_CONNECT_TIMEOUT
  connect_timeout = Max(user_specified_timeout, MIN_CONNECT_TIMEOUT)
  current_backoff = INITIAL_BACKOFF

  // Make a connection attempt with the given timeout
  while (TryConnect(connect_timeout) != SUCCESS)

    // Once the connection attempt has failed, then backoff before trying again
    current_deadline = now() + current_backoff
    SleepUntil(current_deadline)
    current_backoff = Min(current_backoff * MULTIPLIER + UniformRandom(-JITTER * current_backoff, JITTER * current_backoff), MAX_BACKOFF)
```

Where the backoff and timeout are decoupled, and backoff is actual idle time between connection attempts. I can't speak to whether this is better or worse, but it's what I expected when setting the parameter.

I would say it's safe to close this issue at this point.

**Comment by dfawley** [comment]
> The question in this particular bug is what deadline to use for whatever handshaking has to be done after the TCP connection has been established.

I think the spec isn't clear.  Apparently Go's interpretation of it was that "TryConnect" _included_ the TLS and HTTP/2 handshaking steps as well, while it sounds like you're saying C is only counting the TCP handshake.

**Comment by ejona86** [comment]
Java teams' understanding of TryConnect is that it includes TLS and HTTP/2. Note that otherwise you would need another timeout in place (which we'd all agree there is no such other timeout defined). Using the deadline from TryConnect allows you to get the auto-increasing timeout that allows connecting in pathological conditions. But I will note that the proper timeout handling [is not implemented in Java](https://github.com/grpc/grpc-java/issues/1943).

@alanctgardner, if you view the algorithm's backoff from the server's perspective, you'll generally see normal backoff behavior. So coupling backoff and timeout decreases time to retry without removing the benefit of the backoff. Also, if the connection attempt is timing out, there was likely no/minimal progress the majority of the time (it's effectively sleeping waiting for retransmits and the like).

**Comment by markdroth** [comment]
I agree that the spec for this is not entirely clear.  We will at some point in the future write a new spec for all aspects of client channel behavior, including reconnection backoff, connectivity state semantics, and resolver and LB policy behavior.  But for now, let's do the best we can with what we've got.

C-core does actually include TLS handshaking in the connection attempt.  It does *not* currently include the HTTP/2 settings exchange, but that's something we can fix separately as part of addressing the DoS concern.

With regard to this bug, after discussing this in detail with Doug and Eric yesterday, I think we are indeed doing this wrong.  The problem with our current code is that the backoff deadline is being used for both the next time we should start a retry and the deadline for the current attempt, but these should actually be two different numbers: the deadline of a given connection attempt can be larger than the backoff deadline.

For example, for the first connection attempt, the backoff is 1 second, but the minimum timeout is 20 seconds.  What that means is that if the first connection attempt fails in less than 1 second, we should until 1 second has elapsed before trying again.  However, if the connection attempt is still in progress after 1 second, we should let it continue until either it finishes or 20 seconds have elapsed, at which point we immediately make the next attempt.

Given that, I think the fix needs to be linked to David's work in #12867 after all.  His initially proposed solution for that bug will actually make the problem described in this bug worse if we don't fix both at the same time, because it will lower the initial backoff deadline to 1 second and continue to use that as the connection deadline as well.  Instead, we'll need to change our code to understand both numbers independently.

Reassigning to David to work on a fix for both bugs.

**Comment by ctiller** [comment]
Hoping we end up somewhere like:
- next retry attempt is the initial deadline
- we guarantee not to time out for at least 20 seconds after any progress
towards connecting is made

**Comment by markdroth** [comment]
We can probably do that as a future optimization.  Right now, let's focus on fixing our existing broken behavior.

**Comment by ctiller** [comment]
I mention it because not moving the timeout is broken behavior if we're
making progress, and the implementation ought to be trivial:

Expose a grpc_millis for the last activity to each component (this is also
helpful a gpr_atm). Store now blindly to it whenever we make progress.

At connection start time, start a timer with our initial guess. When it
expires check if now >= 20+last activity, if it is about the attempt. If
not, start a new timer to expire at that time.

**Comment by dfawley** [comment]
> I mention it because not moving the timeout is broken behavior if we're making progress, and the implementation ought to be trivial:

This assumes that activity is an indication of making progress, and I don't believe that's necessarily true, because we are running user handshaking code for at least one phase of this.  I think it would be more appropriate to set a deadline on the full connection establishment so that if the handshaking is confused and can't properly complete (but still generates traffic), it will still eventually terminate.  Yes that's unlikely, but any potential for an unbounded operation worries me.

**Comment by dgquintas** [comment]
@markdroth , regarding "we should let it continue until either it finishes or 20 seconds have elapsed, at which point we immediately make the next attempt.", that's not the way the spec currently describes the behavior. I presume you are suggesting the following:

```
ConnectWithBackoff() {
  current_backoff = INITIAL_BACKOFF
  current_deadline = now() + INITIAL_BACKOFF

  // TryConnect returns {SUCCESS, TIMEOUT, FAILED}
  while ( 
    (connection_result = TryConnect(Max(current_deadline, now() + MIN_CONNECT_TIMEOUT)))
       != SUCCESS) {
    if (connection_result == FAILED) {
      SleepUntil(current_deadline)
      current_backoff = Min(current_backoff * MULTIPLIER, MAX_BACKOFF)
      current_deadline = now() + current_backoff +
        UniformRandom(-JITTER * current_backoff, JITTER * current_backoff)
    }  
  }
}
```

That is to say, we only backoff upon **explicit failures**, not _timeouts_. 

@markdroth does this describe your intent?

**Comment by markdroth** [comment]
No, I did not mean that we should differentiate between explicit failures and timeouts; that difference should be irrelevant to the backoff calculation.  I think the pseudo-code in the spec is fine as-is.

What I mean is that, because the deadline for the current connection attempt may be longer than the backoff, if the current attempt fails after taking longer than the backoff amount, we should retry immediately.

In other words, in the pseudo-code shown in the spec, the SleepUntil() will be a no-op if current_deadline is in the past when the attempt completes.

I think the problem in our code is that we actually need two different values, current_backoff and current_deadline, and our code is currently structured to use only a single value for both.  I think that's the part that needs to change.

**Comment by markdroth** [comment]
Doug, I'm not actually worried about that case, because if the handshaking code is actively talking but not making progress, there's no reason to think that it will behave any differently on the next attempt.  So in principle, I agree with Craig that it should be preferable to defer the timeout if we are making progress.

That having been said, I would prefer not to make this change now, for two reasons.  First, I think will be more invasive, because it will require changes to the handshaker API.  And second, I would prefer to take more time to think through the exact parameters and make sure that there are no other implications before we implement this.  So for now, let's hold off on that change.  However, I will add a note about this in my client channel roadmap doc, so we can tackle it later (perhaps as part of writing up a comprehensive spec for client channel behavior).

**Comment by ctiller** [comment]
Sounds good. A small nuance I think was missed and worth noting: the
timestamp should be updated when progress is made. This is different to
receiving or sending bytes: I'd lean towards only updating on state
transition boundaries, and possibly not even all of them. Such a scheme
would (I believe) be robust against the failure case Doug is concerned with
also.

**Comment by markdroth** [comment]
Ah, I see -- so instead of updating the timestamp when we send or receive bytes, we would instead update it when (e.g.) an individual handshaker completes.  That makes sense.

Anyway, I've added this to my doc, so we can consider this further when we get a chance to look at this in more detail.

**Comment by dgquintas** [comment]
Fixed by https://github.com/grpc/grpc/pull/13121
